toolchain/check/handle_impl.cpp

```diff
@@ -99,7 +107,36 @@ static auto ExtendImpl(Context& context, Parse::AnyImplDeclId parse_node,
     context.emitter().Emit(parse_node, ExtendImplOutsideClass);
     return;
   }
-  auto& enclosing_scope = context.name_scopes().Get(enclosing_scope_id);
+
+  if (params_node.is_valid()) {
+    CARBON_DIAGNOSTIC(ExtendImplForall, Error,
+                      "Cannot `extend` a parameterized `impl`.");
+    context.emitter().Emit(extend_node, ExtendImplForall);
+    enclosing_scope.has_error = true;
+    return;
+  }
+
+  if (context.parse_tree().node_kind(self_type_node) ==
+      Parse::NodeKind::TypeImplAs) {
+    CARBON_DIAGNOSTIC(ExtendImplSelfAs, Error,
+                      "Cannot `extend` an `impl` with an explicit self type.");
+    CARBON_DIAGNOSTIC(ExtendImplSelfAsDefault, Note,
+                      "Remove the explicit `Self` type here.");
```

Comment:
FWIW I kind of lean towards the "declare variables close to where they're used" in cases like this, due to the two-scope separation. (it means I wouldn't be looking to see if there are other spots this diagnostic is produced)

Reply:
Done.